Fixes #660: Improve typing/logic for `options` in decode, decode_complete; Improve docs

pachewise opened the pull request:
Basically trying to remove "Unknown" or "partially unknown" pylance issues. Mostly with `options`, but there were a few simple other fixes (kwargs, segments, etc). This showed some minor bugs and a few code smells (passing too many args to `api_jws.decode_complete`, for example).

- [x] We should include Options / SigOptions / FullOptions in the docs somehow
- [x] `PyJWT()` is used as a singleton. Do we want to keep the `options` param in that case? Wondering because codecov is warning that we don't have a test for `options is not None` case.

pachewise commented on jwt/api_jwt.py:
```diff
@@ -236,9 +261,9 @@ def decode(
-        issuer=None,
+        options: FullOptions,
+        audience: Iterable[str] | str | None = None,
+        issuer: Iterable[str] | str | None = None,
```
```suggestion
        issuer: Container[str] | str | None = None,
```
see discussion in #1040

auvipy reviewed (commented):
ping me when it is ready for review

pachewise commented:
@auvipy this should be ready to review. Note that I changed the documentation to use docstrings/autodoc so that the docs don't get stale (I saw a few cases where that happened). Also added more docs where needed; let me know if you'd rather me separate that into another PR. 

I still need an answer on whether we want to support the `PyJWT(options)` call. If no opinion, I will leave it untouched for now and add a test for the missing branch (see codecov failure)

pachewise commented on CHANGELOG.rst:
```diff
@@ -11,7 +11,7 @@ Fixed
 - Validate key against allowed types for Algorithm family in `#964 <https://github.com/jpadilla/pyjwt/pull/964>`__
 - Add iterator for JWKSet in `#1041 <https://github.com/jpadilla/pyjwt/pull/1041>`__
-
+- Improve typing/logic for `options` in decode, decode_complete by @pachewise in `#1045 <https://github.com/jpadilla/pyjwt/pull/1045>`__
```
In Added section, include that we added docs for Algorithm, types, etc.

auvipy reviewed (changes requested):
can you please fix the merge conflicts?
